registry/storage/garbagecollect.go

```diff
@@ -109,42 +121,61 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 		}
 
 		err = manifestEnumerator.Enumerate(ctx, func(dgst digest.Digest) error {
-			removeDueToUntagged := false
-			removeDueToNotNeeded := false
-
-			if exhaustiveNeeded != nil && len(*exhaustiveNeeded) > 0 {
-				// For this repo we know exhaustively which images are needed.
-				// If this one is not part of that, get rid of it :-)
-				if _, needed := (*exhaustiveNeeded)[dgst]; !needed {
-					removeDueToNotNeeded = true
-					emit("remove manifest %s: not needed", dgst)
+			if opts.DeleteManifests {
+				modTime, ok := blobModTime[dgst]
+
+				if !ok {
+					// in this case it was probably uploaded whilst the run started -> we will skip deleting it
+					emit("manifest %s does not have a mod time", dgst)
 				}
-			}
 
-			if !removeDueToNotNeeded && opts.RemoveUntagged {
-				// fetch all tags where this manifest is the latest one
-				tags, _, err := tagsService.Lookup2(ctx, distribution.Descriptor{Digest: dgst})
-				if err != nil {
-					return fmt.Errorf("failed to retrieve tags for digest %v: %v", dgst, err)
+				if ok && modTime.Before(opts.OlderThan) {
+					// Get tags that currently refer to this manifest and those that used to refer to it.
```

Comment:
let's add something like
						emit("remove manifest %s: %s is older than %s", dgst, modtime, olderthanduration)